ksml-integration-tests/src/test/java/io/axual/ksml/integration/BinaryDataFormatIT.java

```diff
@@ -0,0 +1,239 @@
+
+        // Wait for KSML to be ready
+        waitForKSMLReady();
+    }
+
+    @Test
+    @Timeout(90) // 1.5 minutes should be enough for binary processing
+    void testKSMLBinaryProcessing() throws Exception {
+        // Wait for first binary data to be generated and processed
+        log.info("Waiting for KSML to generate and process binary data...");
+        waitForBinaryDataGeneration();
+
+        // Verify KSML is still running
+        assertThat(ksmlRunner.isRunning()).isTrue().as("KSMLRunner should still be running");
+
+        // Create consumer properties
+        Properties consumerProps = new Properties();
+        consumerProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka.getBootstrapServers());
+        consumerProps.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        consumerProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+        consumerProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+
+        // Check ksml_sensordata_binary topic (producer output - binary data)
+        consumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, "test-consumer-binary");
+        Map<String, byte[]> originalMessages = new HashMap<>();
+        try (KafkaConsumer<String, byte[]> consumer = new KafkaConsumer<>(consumerProps)) {
+            consumer.subscribe(Collections.singletonList("ksml_sensordata_binary"));
+            ConsumerRecords<String, byte[]> records = KSMLRunnerTestUtil.pollWithRetry(consumer, Duration.ofSeconds(10));
+
+            assertThat(records).isNotEmpty().as("Should have generated binary data in ksml_sensordata_binary topic");
+            log.info("Found {} binary messages", records.count());
+
+            // Validate binary messages and store for comparison
+            records.forEach(record -> {
+                log.info("Binary message: key={}, bytes={}", record.key(), Arrays.toString(record.value()));
+                assertThat(record.key()).startsWith("msg").as("Message key should start with 'msg'");
+
+                // Validate binary structure - should be 7 bytes
+                byte[] binaryValue = record.value();
+                assertThat(binaryValue).hasSize(7).as("Binary message should have 7 bytes");
+
+                // Check that bytes 2-5 are ASCII 'KSML' (K=75, S=83, M=77, L=76)
+                assertThat(binaryValue[2] & 0xFF).isEqualTo(75).as("Third byte should be ASCII 'K' (75)");
+                assertThat(binaryValue[3] & 0xFF).isEqualTo(83).as("Fourth byte should be ASCII 'S' (83)");
+                assertThat(binaryValue[4] & 0xFF).isEqualTo(77).as("Fifth byte should be ASCII 'M' (77)");
+                assertThat(binaryValue[5] & 0xFF).isEqualTo(76).as("Sixth byte should be ASCII 'L' (76)");
+
+                // Store original message for later comparison
+                originalMessages.put(record.key(), Arrays.copyOf(binaryValue, binaryValue.length));
```

Comment:
By using a HashMap you are losing the order in which the messages are produced. Shouldn't order checks also be part of the tests?